src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```
@@ -54,9 +54,22 @@ internal Utf8JsonStreamReader(Stream stream, int bufferSize = BufferSizeDefault,
             _buffer = _bufferPool.Rent(bufferSize);
             _disposed = false;
             _stream = stream;
-#pragma warning disable CA2022 // Avoid inexact read
-            _stream.Read(_buffer, 0, 3);
-#pragma warning restore CA2022
+#if NET
+            _stream.ReadExactly(_buffer, 0, 3);
```

Hi @nkolev92 , @zivkan , I'm not sure who is the owner of the code, but I saw that you did a code review recently. I have a problem and need your help. I'm fixing the CA2022 warnings by calling `Stream.ReadExactly`, but this causes all unit tests to fail with: `System.IO.EndOfStreamException: Unable to read beyond the end of the stream.` The test data contains extra bytes and fails on last 3 bytes. Can I change the count to another number? Or do you have any other suggestions?

---

Isn't a difference in behavior between ReadExactly and Read? 

We're basically trying to read the BOM if it exists, and it may not always exist.

I don't know if ReadExactly is appropriate here.

---

Yes, the behavior between ReadExactly and Read is different. `ReadExactly` requires exactly N bytes to continue and will throw `EndOfStreamException` if it can't read exactly the requested number of bytes.

I've confirmed that whether using `ReadExactly` or `Read`, the data read into the buffer remains the same (even though ReadExactly throws EndOfStreamException).

Is the purpose of this code to compare the first three bytes in the buffer (regardless of whether a BOM exist) with UTF-8 BOM signature? If yes, I will add try-catch method to get the bytes.

---

@jgonz120 can you help here? 

I think so, but Jonatan wrote it and he'll know for sure :)

---

Hi @jgonz120 , could you please confirm if we can replace `Read` with `ReadExaclty` that includes a try-catch here?

---

> The test data contains extra bytes and fails on last 3 bytes.

Are you sure it's failing on the last 3 bytes? I would expect this code to run only on the first bytes of a stream.

Reading the code, the intent appears to be to skip the UTF8 byte order marker (BOM) if the BOM exists at the beginning of the stream, and appears to be copied from [System.Text.Json's docs](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-utf8jsonreader#filter-data-using-utf8jsonreader). But the BOM is optional, not all valid json streams will have it. The UTF8 BOM is 3 bytes long, hence trying to read 3 bytes, but `{}`, `[]`, and `""` are all valid two byte json strings. Hence it's incorrect to assume that any incoming stream will be at least 3 bytes long.

So, if the stream contains only 2 bytes, we know that it doesn't contain the UTF8 BOM, and we can ask System.Text.Json to try to parse the data.  For this reason, I think `ReadExactly` is the wrong API for this method.

---

>Are you sure it's failing on the last 3 bytes? I would expect this code to run only on the first bytes of a stream.

Sorry that I might have misspoken in my explanation. `ReadExactly` requires exactly N bytes to continue and will throw `EndOfStreamException` if it can't read exactly the requested number of bytes.
The first failure occurs due to the hardcode 3 bytes read requrement, the stream's remaining bytes are fewer than 3 in the final read, and if the stream contains only 2 bytes, `ReadExactly` will also fail with `EndOfStreamException`.
https://github.com/NuGet/NuGet.Client/blob/ec1f498d0bf5429ea23ad42c564e8927516d57ce/src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs#L60

I add try-catch to handle the `EndOfStreamException` then we can get all bytes normally.
https://github.com/NuGet/NuGet.Client/blob/ec1f498d0bf5429ea23ad42c564e8927516d57ce/src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs#L57-L78

In fact, `ReadExactly` will call `read` method except count is zero, so the data read into the buffer remains the same (even though ReadExactly throws EndOfStreamException) whether using `ReadExactly` or `Read`.
https://github.com/dotnet/runtime/blob/7cc32192cbb4894bfebbb09a903215993207a3c0/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L845-L850
https://github.com/dotnet/runtime/blob/7cc32192cbb4894bfebbb09a903215993207a3c0/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L886-L908

Kindly confirm whether `ReadExactly` remains an inappropriate choice for this method. I'll revert my changes if that's the case.

---

Hi @zivkan , could you please verify again if `ReadExactly` with try-catch is an appropriate replacement for `Read` here? If you maintain that `ReadExactly` is the wrong API for this method, I'll revert this change.

---

My gut reaction is how to handle the case when there is fewer than 3 bytes read. Ignoring an exception means we can't know how many bytes were actually read. The code then hardcodes checking the first 3 bytes of the array, which means it checks bytes that might not have been written and therefore is uninitialized.

However, the old code also didn't handle this. So, in my opinion , while it's technically "wrong", from the point of view of this PR, it's not making it worse. The problem already existed.

@jgonz120 what do you think?

---

Update: https://github.com/NuGet/NuGet.Client/blob/e0031a3eff1d81473875cb31f92afa887ddee959/src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs#L58-L73

If an `EndOfStreamException` is caught, then get the actual number of bytes and re-read the stream. Finally, only compare the bytes that were actually read.

---

Hi @zivkan , does this modification make sense to you?

---

Hi @jgonz120 , @zivkan , could you please do another quick code review for the new code? Will this work for the stream contains less than 3 bytes? Does it fit your needs?

---

Hi @zivkan , @jgonz120 , @nkolev92 , just wanted to check in on the code review. Let me know if you had any further thoughts or if there's anything you'd like me to update. Thanks again. :)